File: Glotto.MacOS/Glotto/Glotto/LocalRuleTransliterationProvider.swift
import Foundation

// MARK: - Rule Protocols and Models

private protocol TransliterationRule {
    var latin: String { get }
    var uni: String { get }
}

private struct ConsonantRule: TransliterationRule {
    let latin: String
    let uni: String
}

private struct VowelRule: TransliterationRule {
    let latin: String
    let independent: String
    let modifier: String
    var uni: String { independent }
}

private struct LiteralRule: TransliterationRule {
    let latin: String
    let uni: String
}

private struct GayanukittaRule: TransliterationRule {
    let latin: String
    let uni: String
}

// MARK: - Local Rule Transliteration Provider

/// An offline, rule-based transliterator mapping phonetic Singlish to Sinhala Unicode.
/// Conforms to `TransliterationProvider`.
final class LocalRuleTransliterationProvider: TransliterationProvider, @unchecked Sendable {

    private static let HAL_KIRIMA = "්"

    // MARK: - Static Rule Databases

    private static let VOWELS: [VowelRule] = [
        VowelRule(latin: "aa", independent: "ආ", modifier: "ා"),
        VowelRule(latin: "a)", independent: "ආ", modifier: "ා"),
        VowelRule(latin: "Aa", independent: "ඈ", modifier: "ෑ"),
        VowelRule(latin: "A)", independent: "ඈ", modifier: "ෑ"),
        VowelRule(latin: "ae", independent: "ඈ", modifier: "ෑ"),
        VowelRule(latin: "ii", independent: "ඊ", modifier: "ී"),
        VowelRule(latin: "i)", independent: "ඊ", modifier: "ී"),
        VowelRule(latin: "ie", independent: "ඊ", modifier: "ී"),
        VowelRule(latin: "ee", independent: "ඊ", modifier: "ී"),
        VowelRule(latin: "ea", independent: "ඒ", modifier: "ේ"),
        VowelRule(latin: "e)", independent: "ඒ", modifier: "ේ"),
        VowelRule(latin: "ei", independent: "ඒ", modifier: "ේ"),
        VowelRule(latin: "oo", independent: "ඌ", modifier: "ූ"),
        VowelRule(latin: "uu", independent: "ඌ", modifier: "ූ"),
        VowelRule(latin: "u)", independent: "ඌ", modifier: "ූ"),
        VowelRule(latin: "au", independent: "ඖ", modifier: "ෞ"),
        VowelRule(latin: "a", independent: "අ", modifier: ""),
        VowelRule(latin: "A", independent: "ඇ", modifier: "ැ"),
        VowelRule(latin: "i", independent: "ඉ", modifier: "ි"),
        VowelRule(latin: "e", independent: "එ", modifier: "ෙ"),
        VowelRule(latin: "u", independent: "උ", modifier: "ු"),
        VowelRule(latin: "o", independent: "ඔ", modifier: "ො"),
        VowelRule(latin: "I", independent: "ඓ", modifier: "ෛ")
    ]

    private static let CONSONANTS: [ConsonantRule] = [
        ConsonantRule(latin: "nng", uni: "ඟ"),
        ConsonantRule(latin: "nd", uni: "ඳ"),
        ConsonantRule(latin: "nND", uni: "ඬ"),
        ConsonantRule(latin: "mb", uni: "ඹ"),

        // velar
        ConsonantRule(latin: "k", uni: "ක"),
        ConsonantRule(latin: "kh", uni: "ඛ"),
        ConsonantRule(latin: "g", uni: "ග"),
        ConsonantRule(latin: "gh", uni: "ඝ"),

        // palatal
        ConsonantRule(latin: "ch", uni: "ච"),
        ConsonantRule(latin: "Ch", uni: "ඡ"),
        ConsonantRule(latin: "j", uni: "ජ"),
        ConsonantRule(latin: "q", uni: "ඣ"),
        ConsonantRule(latin: "GN", uni: "ඥ"),
        ConsonantRule(latin: "KN", uni: "ඤ"),

        // retroflex
        ConsonantRule(latin: "T", uni: "ට"),
        ConsonantRule(latin: "Th", uni: "ඨ"),
        ConsonantRule(latin: "D", uni: "ඩ"),
        ConsonantRule(latin: "Dh", uni: "ඪ"),
        ConsonantRule(latin: "N", uni: "ණ"),

        // dental
        ConsonantRule(latin: "t", uni: "ත"),
        ConsonantRule(latin: "th", uni: "ථ"),
        ConsonantRule(latin: "d", uni: "ද"),
        ConsonantRule(latin: "dh", uni: "ධ"),
        ConsonantRule(latin: "n", uni: "න"),

        // labial
        ConsonantRule(latin: "p", uni: "ප"),
        ConsonantRule(latin: "ph", uni: "ඵ"),
        ConsonantRule(latin: "b", uni: "බ"),
        ConsonantRule(latin: "bh", uni: "භ"),
        ConsonantRule(latin: "m", uni: "ම"),

        // semivowels
        ConsonantRule(latin: "Y", uni: "ය"),
        ConsonantRule(latin: "y", uni: "ය"),
        ConsonantRule(latin: "r", uni: "ර"),
        ConsonantRule(latin: "l", uni: "ල"),
        ConsonantRule(latin: "L", uni: "ළ"),
        ConsonantRule(latin: "Lu", uni: "ළු"),
        ConsonantRule(latin: "v", uni: "ව"),
        ConsonantRule(latin: "w", uni: "ව"),

        // sibilants / glottal
        ConsonantRule(latin: "sh", uni: "ශ"),
        ConsonantRule(latin: "Sh", uni: "ෂ"),
        ConsonantRule(latin: "s", uni: "ස"),
        ConsonantRule(latin: "h", uni: "හ"),
        ConsonantRule(latin: "f", uni: "ෆ")
    ]

    private static let LITERALS: [LiteralRule] = [
        LiteralRule(latin: "\\n", uni: "ං"),
        LiteralRule(latin: "\\h", uni: "ඃ"),
        LiteralRule(latin: "\\N", uni: "ඞ"),
        LiteralRule(latin: "\\R", uni: "ඍ"),
        LiteralRule(latin: "\\r", uni: "ර්" + "\u{200D}"),
        LiteralRule(latin: "R", uni: "ර්" + "\u{200D}"),
        LiteralRule(latin: "\\y", uni: "ය")
    ]

    private static let GAYANUKITTA: [GayanukittaRule] = [
        GayanukittaRule(latin: "ruu", uni: "ෲ"),
        GayanukittaRule(latin: "ru", uni: "ෘ")
    ]

    // MARK: - Sorted Rules (Precomputed on demand)

    private let vowelsByLength: [VowelRule]
    private let consonantsByLength: [ConsonantRule]
    private let literalsByLength: [LiteralRule]
    private let gayanukittaByLength: [GayanukittaRule]

    // MARK: - Init

    init() {
        self.vowelsByLength = Self.VOWELS.sorted { $0.latin.count > $1.latin.count }
        self.consonantsByLength = Self.CONSONANTS.sorted { $0.latin.count > $1.latin.count }
        self.literalsByLength = Self.LITERALS.sorted { $0.latin.count > $1.latin.count }
        self.gayanukittaByLength = Self.GAYANUKITTA.sorted { $0.latin.count > $1.latin.count }
    }

    // MARK: - TransliterationProvider Interface

    func candidates(for text: String, profile: LanguageProfile) async throws -> [TransliterationCandidate] {
        // This local provider strictly supports Sinhala transliteration.
        guard profile.id == "si", !text.isEmpty else { return [] }

        let converted = convert(text: text)
        return [TransliterationCandidate(text: converted, rank: 0)]
    }

    // MARK: - Core Conversion Logic

    private func convert(text: String) -> String {
        // Split by hyphen to support disambiguation boundaries.
        let segments = text.components(separatedBy: "-")
        return segments.map(convertSegment).joined()
    }

    private func convertSegment(_ segment: String) -> String {
        var out = ""
        var i = segment.startIndex

        while i < segment.endIndex {
            // 1. Literal escapes (anusvara, visarga, repaya, yansaya escape, ...)
            if let lit = matchAt(rules: literalsByLength, text: segment, index: i) {
                out.append(lit.uni)
                i = segment.index(i, offsetBy: lit.latin.count)
                continue
            }

            // 2. Consonant-led tokens
            if let cons = matchAt(rules: consonantsByLength, text: segment, index: i) {
                let after = segment.index(i, offsetBy: cons.latin.count)

                // 2a. consonant + gayanukitta (ru / ruu) -> vocalic-r vowel signs
                if let gaya = matchAt(rules: gayanukittaByLength, text: segment, index: after) {
                    out.append(cons.uni + gaya.uni)
                    i = segment.index(after, offsetBy: gaya.latin.count)
                    continue
                }

                // 2b. consonant + "r" (+ optional vowel) -> rakaransaya cluster (ක්ර)
                if after < segment.endIndex && segment[after] == "r" {
                    let next = segment.index(after, offsetBy: 1)
                    if let vowel = matchAt(rules: vowelsByLength, text: segment, index: next) {
                        out.append(cons.uni + Self.HAL_KIRIMA + "\u{200D}" + "ර" + vowel.modifier)
                        i = segment.index(next, offsetBy: vowel.latin.count)
                    } else {
                        out.append(cons.uni + Self.HAL_KIRIMA + "\u{200D}" + "ර")
                        i = next
                    }
                    continue
                }

                // 2c. consonant + vowel
                if let vowel = matchAt(rules: vowelsByLength, text: segment, index: after) {
                    out.append(cons.uni + vowel.modifier)
                    i = segment.index(after, offsetBy: vowel.latin.count)
                    continue
                }

                // 2d. bare consonant, no vowel follows -> hal kirima (vowel-killer)
                out.append(cons.uni + Self.HAL_KIRIMA)
                i = after
                continue
            }

            // 3. Standalone vowel (syllable-initial, no preceding consonant)
            if let vowel = matchAt(rules: vowelsByLength, text: segment, index: i) {
                out.append(vowel.independent)
                i = segment.index(i, offsetBy: vowel.latin.count)
                continue
            }

            // 4. Fallback: already-Sinhala characters, punctuation, spaces, etc.
            out.append(segment[i])
            i = segment.index(after: i)
        }

        return out
    }

    // MARK: - Substring Matching Helper

    private func matchAt<T: TransliterationRule>(rules: [T], text: String, index: String.Index) -> T? {
        for rule in rules {
            if let limit = text.index(index, offsetBy: rule.latin.count, limitedBy: text.endIndex),
               text[index..<limit] == rule.latin {
                return rule
            }
        }
        return nil
    }
}

File: Glotto.WinUI/Glotto.WinUI/Providers/LocalRuleTransliterationProvider.cs
// LocalRuleTransliterationProvider.cs
// Port of LocalRuleTransliterationProvider.swift.
//
// Offline, rule-based transliterator mapping phonetic Singlish to Sinhala Unicode.
// Contains the exact same character tables as the Swift version.
// Longest-match-first algorithm: rules sorted descending by latin length at construction time.

using Glotto.WinUI.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Glotto.WinUI.Providers;

public sealed class LocalRuleTransliterationProvider : ITransliterationProvider
{
    private const string HalKirima = "්";
    private const char ZeroWidthJoiner = '\u200D';

    // MARK: - Rule types

    private readonly record struct VowelRule(string Latin, string Independent, string Modifier);
    private readonly record struct ConsonantRule(string Latin, string Uni);
    private readonly record struct LiteralRule(string Latin, string Uni);
    private readonly record struct GayanukittaRule(string Latin, string Uni);

    // MARK: - Rule tables (identical characters to LocalRuleTransliterationProvider.swift)

    private static readonly VowelRule[] s_vowels =
    [
        new("aa", "ආ", "ා"),
        new("a)", "ආ", "ා"),
        new("Aa", "ඈ", "ෑ"),
        new("A)", "ඈ", "ෑ"),
        new("ae", "ඈ", "ෑ"),
        new("ii", "ඊ", "ී"),
        new("i)", "ඊ", "ී"),
        new("ie", "ඊ", "ී"),
        new("ee", "ඊ", "ී"),
        new("ea", "ඒ", "ේ"),
        new("e)", "ඒ", "ේ"),
        new("ei", "ඒ", "ේ"),
        new("oo", "ඌ", "ූ"),
        new("uu", "ඌ", "ූ"),
        new("u)", "ඌ", "ූ"),
        new("au", "ඖ", "ෞ"),
        new("a",  "අ", ""),
        new("A",  "ඇ", "ැ"),
        new("i",  "ඉ", "ි"),
        new("e",  "එ", "ෙ"),
        new("u",  "උ", "ු"),
        new("o",  "ඔ", "ො"),
        new("I",  "ඓ", "ෛ")
    ];

    private static readonly ConsonantRule[] s_consonants =
    [
        // Prenasalized (longer patterns must come before shorter ones)
        new("nng", "ඟ"),
        new("nd",  "ඳ"),
        new("nND", "ඬ"),
        new("mb",  "ඹ"),

        // Velar
        new("k",  "ක"),
        new("kh", "ඛ"),
        new("g",  "ග"),
        new("gh", "ඝ"),

        // Palatal
        new("ch", "ච"),
        new("Ch", "ඡ"),
        new("j",  "ජ"),
        new("q",  "ඣ"),
        new("GN", "ඥ"),
        new("KN", "ඤ"),

        // Retroflex
        new("T",  "ට"),
        new("Th", "ඨ"),
        new("D",  "ඩ"),
        new("Dh", "ඪ"),
        new("N",  "ණ"),

        // Dental
        new("t",  "ත"),
        new("th", "ථ"),
        new("d",  "ද"),
        new("dh", "ධ"),
        new("n",  "න"),

        // Labial
        new("p",  "ප"),
        new("ph", "ඵ"),
        new("b",  "බ"),
        new("bh", "භ"),
        new("m",  "ම"),

        // Semivowels
        new("Y",  "ය"),
        new("y",  "ය"),
        new("r",  "ර"),
        new("l",  "ල"),
        new("L",  "ළ"),
        new("Lu", "ළු"),
        new("v",  "ව"),
        new("w",  "ව"),

        // Sibilants / glottal
        new("sh", "ශ"),
        new("Sh", "ෂ"),
        new("s",  "ස"),
        new("h",  "හ"),
        new("f",  "ෆ")
    ];

    private static readonly LiteralRule[] s_literals =
    [
        new(@"\n", "ං"),
        new(@"\h", "ඃ"),
        new(@"\N", "ඞ"),
        new(@"\R", "ඍ"),
        new(@"\r", "ර්" + ZeroWidthJoiner),
        new("R",   "ර්" + ZeroWidthJoiner),
        new(@"\y", "ය")
    ];

    private static readonly GayanukittaRule[] s_gayanukitta =
    [
        new("ruu", "ෲ"),
        new("ru",  "ෘ")
    ];

    // MARK: - Pre-sorted rule arrays (longest-match-first — critical correctness guarantee)

    private readonly VowelRule[]       _vowelsByLength;
    private readonly ConsonantRule[]   _consonantsByLength;
    private readonly LiteralRule[]     _literalsByLength;
    private readonly GayanukittaRule[] _gayanukittaByLength;

    public LocalRuleTransliterationProvider()
    {
        _vowelsByLength      = [.. s_vowels.OrderByDescending(r => r.Latin.Length)];
        _consonantsByLength  = [.. s_consonants.OrderByDescending(r => r.Latin.Length)];
        _literalsByLength    = [.. s_literals.OrderByDescending(r => r.Latin.Length)];
        _gayanukittaByLength = [.. s_gayanukitta.OrderByDescending(r => r.Latin.Length)];
    }

    // MARK: - ITransliterationProvider

    public Task<IReadOnlyList<TransliterationCandidate>> GetCandidatesAsync(
        string text,
        LanguageProfile profile,
        CancellationToken cancellationToken)
    {
        // This local provider strictly supports Sinhala transliteration.
        if (profile.Id != "si" || string.IsNullOrEmpty(text))
            return Task.FromResult<IReadOnlyList<TransliterationCandidate>>([]);

        var converted = Convert(text);
        return Task.FromResult<IReadOnlyList<TransliterationCandidate>>(
        [
            new TransliterationCandidate(converted, 0)
        ]);
    }

    // MARK: - Core conversion logic

    private string Convert(string text)
    {
        // Split by hyphen to support disambiguation boundaries (same as Swift implementation).
        var segments = text.Split('-');
        return string.Concat(segments.Select(ConvertSegment));
    }

    private string ConvertSegment(string segment)
    {
        var sb = new System.Text.StringBuilder(segment.Length * 2);
        var i = 0;

        while (i < segment.Length)
        {
            // 1. Literal escapes (anusvara, visarga, repaya, yansaya, ...)
            if (MatchLiteral(segment, i) is { } lit)
            {
                sb.Append(lit.Uni);
                i += lit.Latin.Length;
                continue;
            }

            // 2. Consonant-led tokens
            if (MatchConsonant(segment, i) is { } cons)
            {
                var after = i + cons.Latin.Length;

                // 2a. consonant + gayanukitta (ru / ruu) → vocalic-r vowel signs
                if (MatchGayanukitta(segment, after) is { } gaya)
                {
                    sb.Append(cons.Uni).Append(gaya.Uni);
                    i = after + gaya.Latin.Length;
                    continue;
                }

                // 2b. consonant + "r" (+ optional vowel) → rakaransaya cluster (ක්ර)
                if (after < segment.Length && segment[after] == 'r')
                {
                    var next = after + 1;
                    if (MatchVowel(segment, next) is { } vowelAfterR)
                    {
                        sb.Append(cons.Uni).Append(HalKirima).Append(ZeroWidthJoiner).Append('ර').Append(vowelAfterR.Modifier);
                        i = next + vowelAfterR.Latin.Length;
                    }
                    else
                    {
                        sb.Append(cons.Uni).Append(HalKirima).Append(ZeroWidthJoiner).Append('ර');
                        i = next;
                    }
                    continue;
                }

                // 2c. consonant + vowel
                if (MatchVowel(segment, after) is { } vowel)
                {
                    sb.Append(cons.Uni).Append(vowel.Modifier);
                    i = after + vowel.Latin.Length;
                    continue;
                }

                // 2d. bare consonant, no vowel follows → hal kirima (vowel-killer)
                sb.Append(cons.Uni).Append(HalKirima);
                i = after;
                continue;
            }

            // 3. Standalone vowel (syllable-initial, no preceding consonant)
            if (MatchVowel(segment, i) is { } standaloneVowel)
            {
                sb.Append(standaloneVowel.Independent);
                i += standaloneVowel.Latin.Length;
                continue;
            }

            // 4. Fallback: already-Sinhala characters, punctuation, spaces, etc.
            sb.Append(segment[i]);
            i++;
        }

        return sb.ToString();
    }

    // MARK: - Typed match helpers (avoids boxing/reflection)

    private VowelRule? MatchVowel(string text, int index)
    {
        foreach (ref readonly var rule in _vowelsByLength.AsSpan())
        {
            if (index + rule.Latin.Length <= text.Length &&
                text.AsSpan(index, rule.Latin.Length).SequenceEqual(rule.Latin.AsSpan()))
                return rule;
        }
        return null;
    }

    private ConsonantRule? MatchConsonant(string text, int index)
    {
        foreach (ref readonly var rule in _consonantsByLength.AsSpan())
        {
            if (index + rule.Latin.Length <= text.Length &&
                text.AsSpan(index, rule.Latin.Length).SequenceEqual(rule.Latin.AsSpan()))
                return rule;
        }
        return null;
    }

    private LiteralRule? MatchLiteral(string text, int index)
    {
        foreach (ref readonly var rule in _literalsByLength.AsSpan())
        {
            if (index + rule.Latin.Length <= text.Length &&
                text.AsSpan(index, rule.Latin.Length).SequenceEqual(rule.Latin.AsSpan()))
                return rule;
        }
        return null;
    }

    private GayanukittaRule? MatchGayanukitta(string text, int index)
    {
        foreach (ref readonly var rule in _gayanukittaByLength.AsSpan())
        {
            if (index + rule.Latin.Length <= text.Length &&
                text.AsSpan(index, rule.Latin.Length).SequenceEqual(rule.Latin.AsSpan()))
                return rule;
        }
        return null;
    }
}
